File: apps/frontend/src/utils/thinkFilter.ts
const THINK_BLOCK_RE = /<think\b[^>]*>[\s\S]*?<\/think>/gi
const THINK_OPEN_RE = /<think\b[^>]*>[\s\S]*$/gi
const THINK_TAG_RE = /<\/?think\b[^>]*>/gi

export const stripThinkTags = (input: string): string => {
  if (!input) return input

  let cleaned = String(input)
  cleaned = cleaned.replace(THINK_BLOCK_RE, '')
  cleaned = cleaned.replace(THINK_OPEN_RE, '')
  cleaned = cleaned.replace(THINK_TAG_RE, '')
  cleaned = cleaned.replace(/\n{3,}/g, '\n\n')
  return cleaned.trim()
}

export const sanitizeThinkPayload = <T>(value: T): T => {
  if (typeof value === 'string') {
    return stripThinkTags(value) as T
  }
  if (Array.isArray(value)) {
    return value.map((item) => sanitizeThinkPayload(item)) as T
  }
  if (value && typeof value === 'object') {
    const out: Record<string, unknown> = {}
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      out[k] = sanitizeThinkPayload(v)
    }
    return out as T
  }
  return value
}

File: apps/backend/app/utils/think_filter.py
import re
from typing import Any

_THINK_BLOCK_RE = re.compile(r"<think\b[^>]*>[\s\S]*?</think>", re.IGNORECASE)
_THINK_OPEN_RE = re.compile(r"<think\b[^>]*>[\s\S]*$", re.IGNORECASE)
_THINK_TAG_RE = re.compile(r"</?think\b[^>]*>", re.IGNORECASE)


def strip_think_segments(text: str) -> str:
    if not isinstance(text, str) or not text:
        return text

    cleaned = _THINK_BLOCK_RE.sub("", text)
    cleaned = _THINK_OPEN_RE.sub("", cleaned)
    cleaned = _THINK_TAG_RE.sub("", cleaned)
    cleaned = re.sub(r"\n{3,}", "\n\n", cleaned)
    return cleaned.strip()


def sanitize_think_payload(value: Any) -> Any:
    if isinstance(value, str):
        return strip_think_segments(value)
    if isinstance(value, list):
        return [sanitize_think_payload(item) for item in value]
    if isinstance(value, dict):
        return {k: sanitize_think_payload(v) for k, v in value.items()}
    return value
